### PR TITLE
[sql-cli] Implement load file operator

### PR DESCRIPTION
**Pull request by pankajkoti**
# Description

## What is the current behavior?

Currently, we are not having a way for users to load data into databases.

closes: #1227

## What is the new behavior?

Add `load_file` / `LoadFileOperator` to the sql-cli.

## Does this introduce a breaking change?

No.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: python-sdk/src/astro/sql/operators/load_file.py [hunk 1]
```diff
@@ -49,8 +49,11 @@ def __init__(
         native_support_kwargs: dict | None = None,
         columns_names_capitalization: ColumnCapitalization = "original",
         enable_native_fallback: bool | None = LOAD_FILE_ENABLE_NATIVE_FALLBACK,
+        task_id: str | None = None,
         **kwargs,
     ) -> None:
+        self.task_id = task_id or get_unique_task_id("load_file")
+        kwargs["task_id"] = self.task_id
         super().__init__(
             **kwargs_with_datasets(
```
Comment (feluelle) [inline python-sdk/src/astro/sql/operators/load_file.py:58]:
```suggestion
        task_id = task_id or get_unique_task_id("load_file")
        super().__init__(
            task_id=task_id,
            **kwargs_with_datasets(
```
I think we don't need the task_id in `self` or?

File: sql-cli/conftest.py [hunk 2]
```diff
@@ -52,6 +52,7 @@ def sql_file(root_directory, dags_directory):
         root_directory=root_directory,
         path=root_directory / "a.sql",
         target_directory=dags_directory,
+        file_type="sql",
```
Comment (feluelle) [inline sql-cli/conftest.py:55]: This should be set if necessary inside the class `SqlFile`.

File: sql-cli/conftest.py [hunk 3]
```diff
@@ -115,7 +119,7 @@ def sample_dag():
     yield DAG(dag_id, start_date=DEFAULT_DATE)
     with create_session() as session_:
         session_.query(DagRun).delete()
-        session_.query(TI).delete()
+        session_.query(Ti).delete()
```
Comment (feluelle) [inline sql-cli/conftest.py:122]: Why this change?

File: sql-cli/sql_cli/dag_generator.py [hunk 4]
```diff
@@ -7,56 +7,60 @@
...
-from sql_cli.sql_directory_parser import SqlFile, get_sql_files
 from sql_cli.utils.jinja import render
+from sql_cli.workflow_directory_parser import WorkflowFile, get_workflow_files
 
 TEMPLATES_DIRECTORY = Path("templates")
 
 
 @dataclass(frozen=True)
-class SqlFilesDAG:
+class WorkflowFilesDAG:
```
Comment (feluelle) [inline sql-cli/sql_cli/dag_generator.py:17]: How about we call this just `Workflow`?

File: sql-cli/sql_cli/operators/load_file.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+from astro.custom_backend.serializer import deserialize
+from astro.sql import LoadFileOperator
+
+
+def get_load_file_instance(operator_content: dict[str, Any], operator_file_name: str) -> LoadFileOperator:
+    input_file_content = operator_content.pop("input_file")
+    input_file_content["class"] = "File"
+    if "is_dataframe" not in input_file_content:
+        input_file_content["is_dataframe"] = False
```
Comment (feluelle) [inline sql-cli/sql_cli/operators/load_file.py:13]:
```suggestion
    input_file_content.setdefault("is_dataframe", False)
```

Comment (feluelle) [reply sql-cli/sql_cli/operators/load_file.py:13]: See https://docs.python.org/3/library/stdtypes.html#dict.setdefault

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    return {
+        SqlFile(root_directory=directory, path=child, target_directory=target_directory)  # type: ignore
+        for child in directory.rglob(f"*.{SqlFile.file_type}")
+        if child.is_file() and not child.is_symlink()
+    }
+
+
+def get_yaml_files(directory: Path, target_directory: Path | None) -> set[YamlFile]:
+    """
+    Get all sql files within a directory.
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:225]: Please don't forget to update the doc strings.

File: python-sdk/src/astro/sql/operators/load_file.py (hunk 1 above)
Comment (tatiana) [reply python-sdk/src/astro/sql/operators/load_file.py:58]: Yes, I agree, the `BaseOperator` will take care of setting it to self (https://github.com/apache/airflow/blob/main/airflow/models/baseoperator.py#L775)

File: sql-cli/conftest.py [hunk 7]
```diff
@@ -61,6 +62,7 @@ def sql_file_with_parameters(root_directory, dags_directory):
         root_directory=root_directory,
         path=root_directory / "c.sql",
         target_directory=dags_directory,
+        file_type="sql",
```
Comment (tatiana) [inline sql-cli/conftest.py:65]: This feels redundant with the class name. Do we need to expose it?

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+    :param directory: The directory look in for yaml files.
+    :param target_directory: The target directory path for the executable yaml workflow.
+
+    :returns: the workflow files found in the directory.
+    """
+    return {
+        YamlFile(root_directory=directory, path=child, target_directory=target_directory)  # type: ignore
+        for child in directory.rglob(f"*.{YamlFile.file_type}")
+        if child.is_file() and not child.is_symlink()
+    }
```
Comment (tatiana) [inline sql-cli/sql_cli/workflow_directory_parser.py:236]: These two methods are super similar. Would it make sense to unify them into a single `get_files_by_type`?

File: sql-cli/conftest.py (hunk 3 above)
Comment (pankajkoti) [reply sql-cli/conftest.py]: Pycharm complains that camelCase imported as constant, so had made that change. But, reverted now.

Screenshot 2022-11-17 at 12 56 22 PM

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        return TransformOperator(**kwargs)
+
+
+class YamlFile(WorkflowFile):
+    operator_instance_builder_callable_map = {LOAD_FILE_OPERATOR: get_load_file_instance}
+
+    def __init__(self, root_directory: Path, path: Path, target_directory: Path) -> None:
+        super().__init__(root_directory, path, target_directory, YAML_FILE_TYPE)
+        with open(path) as yaml_file:
+            self.yaml_content = yaml.load(yaml_file, Loader=yaml.Loader)
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:188]:
```suggestion
        with path.open() as yaml_file:
            data = yaml.safe_load(yaml_file)
```

Comment (feluelle) [reply sql-cli/sql_cli/workflow_directory_parser.py:188]: Path attributes can be opened using `.open`.

For loading yaml, always use `safe_load`. (see [this](https://stackoverflow.com/a/63916315) for example)

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        self.operator = self._get_operator()
+        self.yaml_file_name = self.path.stem
+
+    def _get_operator(self) -> str:
+        operator = list(self.yaml_content.keys())[0]
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:193]: I think we should validate that only one key is allowed, right? Otherwise the operator returned here could be random as keys is a dict and the order is not guaranteed to be the same.

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+        if operator not in YamlFile.operator_instance_builder_callable_map:
+            raise NotImplementedError(f"Operator support for {operator} not available")
+        return operator
+
+    def to_operator(self) -> LoadFileOperator:
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:198]:
```suggestion
    def to_operator(self) -> BaseOperator:
```
maybe? Or we manually extend e.g. `LoadFileOperator | ExportFileOperator | ...`.

WDYT?

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+            / "/".join(self.get_sub_directories())
+        )
+        target_full_directory.mkdir(parents=True, exist_ok=True)
+
+        target_path = target_full_directory / self.path.name
+
+        target_path.write_text(self.raw_content)
+
+    @abstractmethod
+    def to_operator(self) -> Any:
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:150]:
```suggestion
    def to_operator(self) -> BaseOperator:
```
maybe?

File: sql-cli/sql_cli/constants.py [hunk 13]
```diff
@@ -4,3 +4,11 @@
 DEFAULT_ENVIRONMENT = "default"
 DEFAULT_AIRFLOW_HOME = f".airflow/{DEFAULT_ENVIRONMENT}"
 DEFAULT_DAGS_FOLDER = ".airflow/dags"
+
+GENERATED_WORKFLOW_INCLUDE_DIRECTORY = "include"
+
+SQL_FILE_TYPE = "sql"
+YAML_FILE_TYPE = "yaml"
```
Comment (feluelle) [inline sql-cli/sql_cli/constants.py:11]: We should also allow `yml` files - not only `yaml`.

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+            raise NotImplementedError(f"Operator support for {operator} not available")
+        return operator
+
+    def to_operator(self) -> LoadFileOperator:
+        return self.operator_instance_builder_callable_map[self.operator](
+            self.yaml_content[self.operator], self.yaml_file_name
+        )
+
+
+FILE_TYPE_CLASS_MAP = {SQL_FILE_TYPE: SqlFile, YAML_FILE_TYPE: YamlFile}
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:204]:
```suggestion
SUPPORTED_FILES = {"*.sql": SqlFile, "*.yml|*.yaml": YamlFile}
```
We should have file types together with the file classes - not in separate files. I think this makes it more obvious how we map file types to classes. We can also move `SUPPORTED_FILES` in `constants`.

You could then even have sth like the following to exclude specific yaml config files:
```python
SUPPORTED_FILES = {"*.sql": SqlFile, "[!config]*.yml|[!config]*.yaml": YamlFile}
```

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+        )
+        for child in directory.rglob(f"*.{file_type}")
+        if child.is_file() and not child.is_symlink()
+    }
+
+
+def get_workflow_files(directory: Path, target_directory: Path | None) -> set[WorkflowFile]:
+    sql_files = get_files_by_type(directory, SQL_FILE_TYPE, target_directory=target_directory)
+    yaml_files = get_files_by_type(directory, YAML_FILE_TYPE, target_directory=target_directory)
+    return sql_files.union(yaml_files)
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:231]: Here we should loop over all `SUPPORTED_FILES`.

File: sql-cli/sql_cli/workflow_directory_parser.py (hunk 14 above)
Comment (pankajkoti) [reply sql-cli/sql_cli/workflow_directory_parser.py]: I tried using that pattern, but looks like rglob does not support the pattern you suggested. Added another entry for *.yml in the map.

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+        self.yaml_file_name = self.path.stem
+
+    def _get_operator(self) -> str:
+        top_level_keys = list(self.yaml_content.keys())
+        top_level_keys_count = len(top_level_keys)
+        if top_level_keys_count > 1:
+            raise ScannerError(
+                f"Only one top level operator expected. Got {top_level_keys_count} operators: {top_level_keys}"
+            )
+        operator = list(self.yaml_content.keys())[0]
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:202]:
```suggestion
        operator = top_level_keys[0]
```

File: sql-cli/sql_cli/templates/gen_tasks_dag.py.jinja2 [hunk 17]
```diff
@@ -14,6 +15,10 @@ with DAG(
     start_date=timezone.parse("{{ dag.start_date }}"),
     schedule_interval=None,
 ) as dag:
-    {%- for sql_file in dag.sorted_sql_files() %}
-    {{ define_task(sql_file, dag)|indent(width=4) }}
+    {%- for workflow_file in dag.sorted_workflow_files() %}
+    {% if workflow_file.__class__.__name__ == 'SqlFile'  %}
+    {{ define_transform_task(workflow_file, dag)|indent(width=4) }}
+    {% elif workflow_file.operator == 'load_file'  %}
+    {{ define_load_file_task(workflow_file, dag, workflow_file.get_yaml_content(), workflow_file.yaml_file_name)|indent(width=4) }}
```
Comment (feluelle) [inline sql-cli/sql_cli/templates/gen_tasks_dag.py.jinja2:22]:
```suggestion
    {% elif workflow_file.__class__.__name__ == 'YamlFile'  %}
    {{ define_yaml_task(workflow_file, dag)|indent(width=4) }}
```
Can we make this work^?

Because then we would not need to add this for every operator.

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+            "sql": self.content,
+        }
+        if airflow.__version__.startswith("2.2."):
+            kwargs["op_args"] = []
+        return TransformOperator(**kwargs)
+
+
+class YamlFile(WorkflowFile):
+    operator_instance_builder_callable_map = {LOAD_FILE_OPERATOR: get_load_file_instance}
+    operator_content_callable_map = {LOAD_FILE_OPERATOR: build_load_file_operator_content}
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:186]:
```suggestion
```

File: sql-cli/sql_cli/workflow_directory_parser.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    :param file_type: The type of the workflow file. e.g. sql or yaml
+    :param target_directory: The target directory path for the executable workflow file.
+    """
+
+    def __init__(self, root_directory: Path, path: Path, target_directory: Path, file_type: str) -> None:
+        self.root_directory = root_directory
+        self.path = path
+        self.target_directory = target_directory
+        self.raw_content = self.path.read_text()
+        self.file_type = file_type
```
Comment (feluelle) [inline sql-cli/sql_cli/workflow_directory_parser.py:40]: `WorkflowFile` should not have the file_type if it does not use it. We instantiate different WorkflowFile objects based on the file type already.

File: sql-cli/sql_cli/templates/gen_tasks_dag.py.jinja2 (hunk 17 above)
Comment (pankajkoti) [reply sql-cli/sql_cli/templates/gen_tasks_dag.py.jinja2:22]: I think would be good to keep operator level macros to keep it easier to maintain.

File: sql-cli/include/base/config/default/configuration.yml [hunk 20]
```diff
@@ -7,19 +7,17 @@ connections:
...
-#    login: null
-#    password: null
-#    port: null
-#    schema: null
-
+  - conn_id: bigquery
+    conn_type: bigquery
+    description:
+    extra:
+      project: astronomer-dag-authoring
```
Comment (tatiana) [inline sql-cli/include/base/config/default/configuration.yml:15]: WDYT if we retrieve this from an envvar?

File: sql-cli/sql_cli/macros/load_file_tasks.py.jinja2 [hunk 21]
```diff
@@ -0,0 +1,7 @@
+{% macro define_load_file_task(workflow_file, dag, yaml_content, yaml_file_name) -%}
+from sql_cli.operators.load_file import build_load_file_operator_content
+operator_content = build_load_file_operator_content({{yaml_content|safe}}, '{{yaml_file_name}}')
+{{ workflow_file.get_variable_name() }} = aql.load_file(
+    **operator_content
```
Comment (tatiana) [inline sql-cli/sql_cli/macros/load_file_tasks.py.jinja2:5]: This is much simpler! :raised_hands: did it work as expected? If yes, should we use a similar strategy for the `tranform_task` template?

Comment (pankajkoti) [reply sql-cli/sql_cli/macros/load_file_tasks.py.jinja2]: We would not like to use this approach as the generated DAG is inconsistent to read wrt to Python SQL aql tasks.
@feluelle and I are pairing right not to instead expand the Jinja template to make the generated task look similar to Python SDK example

File: .github/workflows/ci-sql-cli.yaml [hunk 22]
```diff
@@ -100,6 +100,9 @@ jobs:
       AIRFLOW__CORE__XCOM_BACKEND: astro.custom_backend.astro_custom_backend.AstroCustomXcomBackend
       AIRFLOW__ASTRO_SDK__STORE_DATA_LOCAL_DEV: True
       _TYPER_FORCE_DISABLE_TERMINAL: 1
+      GOOGLE_APPLICATION_CREDENTIALS_JSON: ${{ secrets.GOOGLE_APPLICATION_CREDENTIALS_JSON }}
+      GOOGLE_APPLICATION_CREDENTIALS: /tmp/google_credentials.json
```
Comment (pankajkoti) [inline .github/workflows/ci-sql-cli.yaml:104]:
```suggestion
```

File: .github/workflows/ci-sql-cli.yaml [hunk 23]
```diff
@@ -100,6 +101,9 @@ jobs:
...
+      GOOGLE_CLOUD_PORJECT: "astronomer-dag-authoring"
```
Comment (feluelle) [inline .github/workflows/ci-sql-cli.yaml:106]:
```suggestion
```

**Review by feluelle — APPROVED** [review]
LGTM 👍 

Let's only fix the remaining failed SQL CLI test.

All other failing tests are related to python-sdk. :(

File: sql-cli/conftest.py [hunk 24]
```diff
@@ -147,6 +154,38 @@ def initialised_project_with_test_config(initialised_project: Project):
...
+    render(
+        template_file=Path("transform_homes_main.sql.jinja2"),
+        output_file=initialised_project.directory
+        / "workflows"
+        / "example_load_file"
+        / "transform_homes_main.sql",
+        **render_kwargs,
+    )
+
+    return initialised_project
```
Comment (tatiana) [inline sql-cli/conftest.py:186]: It feels that it would be worth having an example workflow with the load. If we had that, would it still make sense to render it in this fixture?

File: sql-cli/include/base/config/default/configuration.yml [hunk 25]
```diff
@@ -7,19 +7,15 @@ connections:
     schema:
     login:
     password:
-
 ## Example of Bigquery connection
-#  - conn_id: bigquery_conn
-#    conn_type: bigquery
-#    description: null
-#    extra:
-#      project: "my-gcp-project"
```
Comment (tatiana) [inline sql-cli/include/base/config/default/configuration.yml:16]: Why did we remove the project from here? Should we adjust the `dev` configuration.yml as well?

File: sql-cli/include/base/config/default/configuration.yml [hunk 26]
```diff
@@ -7,19 +7,15 @@ connections:
...
-#    schema: null
-
+  - conn_id: bigquery
+    conn_type: bigquery
+    description:
+    host:
+    login:
+    password:
+    port:
+    schema:
```
Comment (tatiana) [inline sql-cli/include/base/config/default/configuration.yml:18]: Do we want to have this as a pre-defined connection for the end-users - or should we comment it out - as before? Would this work if they didn't have BQ/GCP providers?

File: sql-cli/pyproject.toml [hunk 27]
```diff
@@ -39,7 +39,7 @@ Jinja2 = ">=2.10.1, <=3.1"
 python-frontmatter = "^1.0.0"
 PyYAML = ">=5.4.1"
 python-dotenv = "^0.21.0"
-astro-sdk-python = {extras = ["all"], version = "^1.2.0"}
+astro-sdk-python = {extras = ["all"], version = "^1.2.2"}
```
Comment (tatiana) [inline sql-cli/pyproject.toml:42]: It would be great to add a comment to why we only support the latest version of the Astro Python SDK

File: sql-cli/sql_cli/operators/load_file.py [hunk 28]
```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
+
+from typing import Any
+
+from astro.custom_backend.serializer import deserialize
+from astro.sql import LoadFileOperator
+
+
+def build_load_file_operator_content(
```
Comment (tatiana) [inline sql-cli/sql_cli/operators/load_file.py:9]: Docstring, please!

File: sql-cli/sql_cli/operators/load_file.py [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+    output_table_content = operator_content.pop("output_table")
+    output_table_content["class"] = "Table"
+    output_table_content.setdefault("name", operator_file_name)
+    output_table_content.setdefault("temp", False)
+    output_table = deserialize(output_table_content)
+    operator_content["output_table"] = output_table
+    return operator_content
+
+
+def get_load_file_instance(yaml_content: dict[str, Any], operator_file_name: str) -> LoadFileOperator:
```
Comment (tatiana) [inline sql-cli/sql_cli/operators/load_file.py:27]: docstring, please

**Review by tatiana — COMMENTED** [review]
Great progress on the load file, @pankajkoti and @feluelle! Left some questions and minor comments

**Comment by feluelle** [comment]
Thanks @tatiana for your feedback!

File: sql-cli/conftest.py (hunk 24 above)
Comment (feluelle) [reply sql-cli/conftest.py:186]: Yes, because what we do in that fixture is that we dynamically change the output table name as when running it concurrently the jobs can interfere on the same table.

File: sql-cli/include/base/config/default/configuration.yml (hunk 25 above)
Comment (feluelle) [reply sql-cli/include/base/config/default/configuration.yml:16]: Because it currently does not work. See https://github.com/astronomer/astro-sdk/issues/1277

File: sql-cli/include/base/config/default/configuration.yml (hunk 26 above)
Comment (feluelle) [reply sql-cli/include/base/config/default/configuration.yml:18]: > Would this work if they didn't have BQ/GCP providers?

We currently install all providers i.e. the user does not have a choice.

File: sql-cli/sql_cli/macros/load_file.py.jinja2 [hunk 30]
```diff
@@ -0,0 +1,58 @@
+{% macro load_file(yaml_file, dag) -%}
+{% set yaml_content = yaml_file.get_yaml_content() -%}
+{{ yaml_file.get_variable_name() }} = aql.load_file(
+    input_file=File(
```
[the remark this answers is not in the document]
Comment (feluelle) [reply sql-cli/sql_cli/macros/load_file.py.jinja2:4]: Thanks @tatiana - good point. I can add a comment there.

File: sql-cli/conftest.py (hunk 24 above)
Comment (pankajkoti) [reply sql-cli/conftest.py:186]: We do have an example workflow with the load that gets added on project initialisation. We added this fixture as tests running concurrently across Airflow version were creating the same output table and were conflicting across different runs of Airflow version. Here, we manipulate the created workflow to have output table with unique name so that concurrent runs succeed well.

File: sql-cli/include/base/config/default/configuration.yml (hunk 25 above)
Comment (pankajkoti) [reply sql-cli/include/base/config/default/configuration.yml:16]: Added a comment in the same ticket to check whether we need to update the dev configuration.yml too.

File: sql-cli/include/base/config/default/configuration.yml (hunk 26 above)
Comment (pankajkoti) [reply sql-cli/include/base/config/default/configuration.yml:18]: Yes, since we install astro-sdk-python[all], the providers would be available.

**Review by feluelle — APPROVED** [review]

File: sql-cli/Makefile [hunk 31]
```diff
@@ -31,7 +31,7 @@ pre-commit:  # Run pre-commit
 flow:  # Build and run the latest version of flow
 
 test:  # Run tests
-	@_TYPER_FORCE_DISABLE_TERMINAL=1 PYTHONWARNINGS="ignore" poetry run pytest -s --cov --cov-branch --cov-report=term-missing tests
+	@_TYPER_FORCE_DISABLE_TERMINAL=1 PYTHONWARNINGS="ignore" poetry run pytest
```
Comment (tatiana) [inline sql-cli/Makefile:34]: @pankajkoti why did we remove the coverage stats?

Comment (pankajkoti) [reply sql-cli/Makefile:34]: @feluelle can we add them back or you have a different opinion?

Comment (feluelle) [reply sql-cli/Makefile:34]: It is now in the project.toml

Comment (pankajkoti) [reply sql-cli/Makefile:34]: I guess Felix moved them to pyproject.toml here: https://github.com/astronomer/astro-sdk/pull/1258/files#diff-f92d64397e38a0a17a38a3b24f90b96020f886a2f8db484165658088325e0c9bR88-R93

Comment (feluelle) [reply sql-cli/Makefile:34]: This way we have the options in all pytest runs - local and in CI.

File: sql-cli/sql_cli/utils/airflow.py [hunk 32]
```diff
@@ -10,7 +10,7 @@
 from packaging.version import Version
 
 if TYPE_CHECKING:
-    from airflow.models.dag import DAG
+    from airflow.models.dag import DAG  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (tatiana) [reply sql-cli/sql_cli/utils/airflow.py:13]: It surprises me it would be a false positive. Did we confirm that tests are going through this if? Could it be the case TYPE_CHECKING is being false?

[the remark this answers is not in the document]
Comment (feluelle) [reply sql-cli/sql_cli/utils/airflow.py:13]: Maybe [this](https://github.com/nedbat/coveragepy/issues/831) helps explaining it.

File: sql-cli/sql_cli/macros/load_file.py.jinja2 (hunk 30 above)
[the remark this answers is not in the document]
Comment (feluelle) [reply sql-cli/sql_cli/macros/load_file.py.jinja2:4]: Completely forgot this. Will add a comment now.

File: sql-cli/conftest.py [hunk 33]
```diff
@@ -147,6 +182,38 @@ def initialised_project_with_test_config(initialised_project: Project):
...
+@pytest.fixture()
+def initialised_project_with_load_file_workflow(
+    root_directory_templates_load_file, initialised_project: Project
+):
+    output_table_name = create_unique_table_name(UNIQUE_HASH_SIZE)
+    render_kwargs = {
+        "context": {"output_table_name": output_table_name},
+        "searchpath": root_directory_templates_load_file,
+    }
+
```
Comment (pankajkoti) [inline sql-cli/conftest.py:194]: Write a comment here on why we need this manipulations.

File: sql-cli/sql_cli/__main__.py [hunk 34]
```diff
@@ -13,10 +13,10 @@
 from sql_cli.astro.command import AstroCommand
 from sql_cli.astro.group import AstroGroup
 from sql_cli.constants import DEFAULT_AIRFLOW_HOME, DEFAULT_DAGS_FOLDER
-from sql_cli.exceptions import ConnectionFailed, DagCycle, EmptyDag, SqlFilesDirectoryNotFound
+from sql_cli.exceptions import ConnectionFailed, DagCycle, EmptyDag, WorkflowFilesDirectoryNotFound
 
 if TYPE_CHECKING:
-    from sql_cli.project import Project
+    from sql_cli.project import Project  # pragma: no cover
```
Comment (pankajkoti) [inline sql-cli/sql_cli/__main__.py:19]: Add a ticket to check on whether we can cover this line and remove the no cover.

File: sql-cli/sql_cli/macros/load_file.py.jinja2 [hunk 35]
```diff
@@ -0,0 +1,64 @@
+{% macro load_file(yaml_file, dag) -%}
+{% set yaml_content = yaml_file.get_yaml_content() -%}
+{{ yaml_file.get_variable_name() }} = aql.load_file(
+    {#- Note: Unfortunately, we cannot simply use **kwargs as some of the deserialized fields
```
Comment (pankajkoti) [inline sql-cli/sql_cli/macros/load_file.py.jinja2:4]: Create a ticket to evaluate if we can support **kwargs with a combination of python helper methods to tranform the yaml content as is needed by Jinja templates.

File: sql-cli/sql_cli/__main__.py (hunk 34 above)
Comment (pankajkoti) [reply sql-cli/sql_cli/__main__.py:19]: Created ticket: https://github.com/astronomer/astro-sdk/issues/1283

File: sql-cli/include/base/config/default/configuration.yml [hunk 36]
```diff
@@ -7,19 +7,15 @@ connections:
...
-#    description: null
-#    extra:
-#      project: "my-gcp-project"
-#    host: null
-#    login: null
-#    password: null
-#    port: null
-#    schema: null
-
+  - conn_id: bigquery
```
[the remark this answers is not in the document]
Comment (pankajkoti) [reply sql-cli/include/base/config/default/configuration.yml:11]: Created ticket: #1282

File: sql-cli/sql_cli/macros/load_file.py.jinja2 (hunk 35 above)
Comment (pankajkoti) [reply sql-cli/sql_cli/macros/load_file.py.jinja2:4]: Created ticket: #1284

**Review by tatiana — APPROVED** [review]
